File: iosApp/iosApp/Theme.swift
// Theme.swift — NEUTRAL base + user-chosen ACCENT (mirrors the Android theme).
// The canvas is neutral (warm charcoal in dark, warm paper in light — NOT green);
// the accent the user picks in Settings (from the shared AccentPalette) recolors
// primary / active / highlight roles in BOTH modes. Nothing is force-green anymore.
//
// Design tokens adapted from the MIT-licensed Hermes Agent web frontend
// (© 2025 Nous Research) — see ATTRIBUTION.md.

import SwiftUI
import Shared

// MARK: - Appearance preference (persisted; dark is the default)

enum ThemeMode: String, CaseIterable, Identifiable {
    case system, dark, light
    var id: String { rawValue }
    var label: String {
        switch self {
        case .system: return "System"
        case .dark: return "Dark"
        case .light: return "Light"
        }
    }
}

// MARK: - Color(hex:) + blend

extension Color {
    init(hex: UInt32) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0,
            opacity: 1
        )
    }

    /// Mix two RGB hexes by ratio r (0 = a, 1 = b) — for accent-tinted surfaces.
    static func blend(_ a: UInt32, _ b: UInt32, _ r: Double) -> Color {
        func mix(_ shift: UInt32) -> Double {
            let av = Double((a >> shift) & 0xFF)
            let bv = Double((b >> shift) & 0xFF)
            return (av + (bv - av) * r) / 255.0
        }
        return Color(.sRGB, red: mix(16), green: mix(8), blue: mix(0), opacity: 1)
    }
}

// MARK: - Semantic theme (mirrors Material's colorScheme roles used by the app)

struct HermesTheme {
    let isDark: Bool

    let background: Color
    let onBackground: Color
    let surface: Color
    let onSurface: Color
    let surfaceVariant: Color
    let onSurfaceVariant: Color
    let secondaryContainer: Color
    let onSecondaryContainer: Color
    let primary: Color
    let onPrimary: Color
    let tertiary: Color
    let onTertiary: Color
    let error: Color
    let onError: Color
    let errorContainer: Color
    let onErrorContainer: Color
    let outline: Color

    // Neutral base hexes (no green identity), matching Android's Color.kt.
    private static let darkBg: UInt32 = 0x121415
    private static let darkSurface: UInt32 = 0x1B1E20
    private static let darkMuted: UInt32 = 0x212528
    private static let darkSecondary: UInt32 = 0x23272A
    private static let lightMuted: UInt32 = 0xECE7DD

    /// Build the active theme from the mode + the user's accent (from Shared).
    static func make(_ mode: ThemeMode, system: ColorScheme, accent: AccentOption) -> HermesTheme {
        let isDark = (mode == .dark) || (mode == .system && system == .dark)
        return isDark ? dark(accent) : light(accent)
    }

    private static func dark(_ accent: AccentOption) -> HermesTheme {
        let a = UInt32(truncatingIfNeeded: accent.darkRgb) & 0xFFFFFF
        let onA = UInt32(truncatingIfNeeded: accent.onDarkRgb) & 0xFFFFFF
        return HermesTheme(
            isDark: true,
            background: Color(hex: darkBg),
            onBackground: Color(hex: 0xECEAE4),
            surface: Color(hex: darkSurface),
            onSurface: Color(hex: 0xECEAE4),
            surfaceVariant: Color(hex: darkMuted),
            onSurfaceVariant: Color(hex: 0xA9ADAB),
            secondaryContainer: Color.blend(darkSecondary, a, 0.26),
            onSecondaryContainer: Color(hex: 0xECEAE4),
            primary: Color(hex: a),
            onPrimary: Color(hex: onA),
            tertiary: Color(hex: a),
            onTertiary: Color(hex: onA),
            error: Color(hex: 0xFB2C36),
            onError: .white,
            errorContainer: Color(hex: 0xFB2C36).opacity(0.15),
            onErrorContainer: Color(hex: 0xFB2C36),
            outline: Color(hex: 0x35393C)
        )
    }

    private static func light(_ accent: AccentOption) -> HermesTheme {
        let a = UInt32(truncatingIfNeeded: accent.lightRgb) & 0xFFFFFF
        let onA = UInt32(truncatingIfNeeded: accent.onLightRgb) & 0xFFFFFF
        return HermesTheme(
            isDark: false,
            background: Color(hex: 0xF6F3EC),
            onBackground: Color(hex: 0x1D1F20),
            surface: Color(hex: 0xFFFFFF),
            onSurface: Color(hex: 0x1D1F20),
            surfaceVariant: Color(hex: lightMuted),
            onSurfaceVariant: Color(hex: 0x5D605F),
            secondaryContainer: Color.blend(lightMuted, a, 0.14),
            onSecondaryContainer: Color(hex: 0x1D1F20),
            primary: Color(hex: a),
            onPrimary: Color(hex: onA),
            tertiary: Color(hex: a),
            onTertiary: Color(hex: onA),
            error: Color(hex: 0xFB2C36),
            onError: .white,
            errorContainer: Color(hex: 0xFB2C36).opacity(0.12),
            onErrorContainer: Color(hex: 0xFB2C36),
            outline: Color(hex: 0xDED8CC)
        )
    }

    /// A neutral default theme for the environment default value (before the accent
    /// loads from the store) — the shared default accent (Blue), inline so it needs
    /// no Kotlin-object access.
    static let dark = HermesTheme(
        isDark: true,
        background: Color(hex: darkBg),
        onBackground: Color(hex: 0xECEAE4),
        surface: Color(hex: darkSurface),
        onSurface: Color(hex: 0xECEAE4),
        surfaceVariant: Color(hex: darkMuted),
        onSurfaceVariant: Color(hex: 0xA9ADAB),
        secondaryContainer: Color.blend(darkSecondary, 0x60A5FA, 0.26),
        onSecondaryContainer: Color(hex: 0xECEAE4),
        primary: Color(hex: 0x60A5FA),
        onPrimary: Color(hex: 0x0A1A2F),
        tertiary: Color(hex: 0x60A5FA),
        onTertiary: Color(hex: 0x0A1A2F),
        error: Color(hex: 0xFB2C36),
        onError: .white,
        errorContainer: Color(hex: 0xFB2C36).opacity(0.15),
        onErrorContainer: Color(hex: 0xFB2C36),
        outline: Color(hex: 0x35393C)
    )
}

// MARK: - Environment plumbing

private struct HermesThemeKey: EnvironmentKey {
    static let defaultValue = HermesTheme.dark
}

extension EnvironmentValues {
    var theme: HermesTheme {
        get { self[HermesThemeKey.self] }
        set { self[HermesThemeKey.self] = newValue }
    }
}

// MARK: - Signature Hermes text treatments (uppercase, wide-tracked "display")

extension Text {
    func hermesDisplayLabel(size: CGFloat = 13) -> Text {
        self.font(.system(size: size, weight: .semibold))
            .tracking(2.2)
    }
}

extension View {
    func hermesMono(size: CGFloat = 12.5) -> some View {
        self.font(.system(size: size, design: .monospaced)).tracking(0.5)
    }
}

File: androidApp/src/main/kotlin/com/personalagent/android/ui/theme/Theme.kt
package com.personalagent.android.ui.theme

import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Shapes
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.personalagent.shared.appearance.AccentOption
import com.personalagent.shared.appearance.AccentPalette

/**
 * Neutral base + user-chosen ACCENT theme. The canvas is neutral (charcoal in
 * dark, warm paper in light — see [Color.kt]); the [AccentOption] the user picks
 * in Settings recolors primary / active / highlight roles in BOTH modes. Nothing
 * is force-green anymore. The signature Hermes *treatments* stay (small radii,
 * uppercase wide-tracked display labels, roomy body leading).
 *
 * Both schemes are built at runtime from the accent, so changing the accent
 * recolors every screen (all screens read `MaterialTheme.colorScheme.*`).
 */

private fun Long.toAccentColor(): Color = Color(0xFF000000L or (this and 0xFFFFFF))

/** Mix [a] toward [b] by [r] (0 = all a, 1 = all b). Used for accent-tinted surfaces. */
private fun blend(a: Color, b: Color, r: Float): Color = Color(
    red = a.red + (b.red - a.red) * r,
    green = a.green + (b.green - a.green) * r,
    blue = a.blue + (b.blue - a.blue) * r,
    alpha = 1f,
)

private fun darkScheme(accent: AccentOption) = run {
    val a = accent.darkRgb.toAccentColor()
    val onA = accent.onDarkRgb.toAccentColor()
    darkColorScheme(
        primary = a,
        onPrimary = onA,
        primaryContainer = blend(NeutralDarkCard, a, 0.24f),
        onPrimaryContainer = NeutralDarkInk,
        secondary = a,
        onSecondary = onA,
        secondaryContainer = blend(NeutralDarkSecondary, a, 0.26f),
        onSecondaryContainer = NeutralDarkInk,
        tertiary = a,
        onTertiary = onA,
        background = NeutralDarkBackground,
        onBackground = NeutralDarkInk,
        surface = NeutralDarkCard,
        onSurface = NeutralDarkInk,
        surfaceVariant = NeutralDarkMuted,
        onSurfaceVariant = NeutralDarkInkDim,
        surfaceContainerLowest = NeutralDarkBackground,
        surfaceContainerLow = NeutralDarkCard,
        surfaceContainer = NeutralDarkCard,
        surfaceContainerHigh = NeutralDarkSecondary,
        surfaceContainerHighest = NeutralDarkSecondary,
        error = StatusDestructive,
        onError = Color(0xFFFFFFFF),
        errorContainer = StatusDestructive.copy(alpha = 0.15f),
        onErrorContainer = StatusDestructive,
        outline = NeutralDarkBorder,
        outlineVariant = NeutralDarkBorder,
        inversePrimary = a,
    )
}

private fun lightScheme(accent: AccentOption) = run {
    val a = accent.lightRgb.toAccentColor()
    val onA = accent.onLightRgb.toAccentColor()
    lightColorScheme(
        primary = a,
        onPrimary = onA,
        primaryContainer = blend(NeutralLightMuted, a, 0.16f),
        onPrimaryContainer = NeutralLightInk,
        secondary = a,
        onSecondary = onA,
        secondaryContainer = blend(NeutralLightMuted, a, 0.14f),
        onSecondaryContainer = NeutralLightInk,
        tertiary = a,
        onTertiary = onA,
        background = NeutralLightBackground,
        onBackground = NeutralLightInk,
        surface = NeutralLightSurface,
        onSurface = NeutralLightInk,
        surfaceVariant = NeutralLightMuted,
        onSurfaceVariant = NeutralLightInkDim,
        surfaceContainerLowest = NeutralLightSurface,
        surfaceContainerLow = NeutralLightBackground,
        surfaceContainer = NeutralLightSurface,
        surfaceContainerHigh = NeutralLightMuted,
        surfaceContainerHighest = NeutralLightMuted,
        error = StatusDestructive,
        onError = Color(0xFFFFFFFF),
        outline = NeutralLightBorder,
        outlineVariant = NeutralLightBorder,
    )
}

/** Persisted appearance preference. Dark is the default mode. */
enum class ThemeMode { SYSTEM, DARK, LIGHT }

// Small radii (0.5rem = 8px default, xl = 12px): crisp, lightly-rounded panels.
private val HermesShapes = Shapes(
    extraSmall = RoundedCornerShape(4.dp),
    small = RoundedCornerShape(6.dp),
    medium = RoundedCornerShape(8.dp),
    large = RoundedCornerShape(12.dp),
    extraLarge = RoundedCornerShape(16.dp),
)

/**
 * The signature Hermes text treatments, reusable across screens.
 *  - [displayLabel]: UPPERCASE, wide letter-spacing, semi-bold — section headers/nav.
 *  - [mono]: tabular monospace for technical/metadata readouts (versions, counts).
 */
object HermesText {
    val displayLabel: TextStyle = TextStyle(
        fontWeight = FontWeight.SemiBold,
        fontSize = 13.sp,
        letterSpacing = 2.2.sp,
        lineHeight = 16.sp,
    )
    val displayLarge: TextStyle = TextStyle(
        fontWeight = FontWeight.Bold,
        fontSize = 20.sp,
        letterSpacing = 1.6.sp,
        lineHeight = 26.sp,
    )
    val mono: TextStyle = TextStyle(
        fontFamily = FontFamily.Monospace,
        fontSize = 12.5.sp,
        letterSpacing = 0.5.sp,
        lineHeight = 18.sp,
    )
}

private val HermesTypography: Typography
    @Composable get() {
        val base = MaterialTheme.typography
        return base.copy(
            bodyLarge = base.bodyLarge.copy(lineHeight = 26.sp, fontSize = 16.sp),
            bodyMedium = base.bodyMedium.copy(lineHeight = 22.sp),
            labelLarge = base.labelLarge.copy(
                fontWeight = FontWeight.Bold,
                letterSpacing = 1.6.sp,
            ),
            titleMedium = base.titleMedium.copy(fontWeight = FontWeight.SemiBold, letterSpacing = 0.4.sp),
            titleSmall = base.titleSmall.copy(fontWeight = FontWeight.SemiBold, letterSpacing = 0.4.sp),
        )
    }

/** Monospace style for code (inline + blocks), sized to sit calmly in body text. */
val CodeTextStyle: TextStyle
    get() = TextStyle(
        fontFamily = FontFamily.Monospace,
        fontSize = 13.5.sp,
        lineHeight = 20.sp,
    )

@Composable
fun PersonalAgentTheme(
    // Dark by default — even when the system is light.
    darkTheme: Boolean = true,
    // The user-selected accent (neutral Blue by default; picked in Settings).
    accent: AccentOption = AccentPalette.DEFAULT,
    content: @Composable () -> Unit,
) {
    MaterialTheme(
        colorScheme = if (darkTheme) darkScheme(accent) else lightScheme(accent),
        typography = HermesTypography,
        shapes = HermesShapes,
        content = content,
    )
}
